Add `keep_decal_opacity` property to allow decals to modify surface alpha

yahkr opened the pull request:
- Closes https://github.com/godotengine/godot-proposals/issues/12245

```glsl
if (decals.data[decal_index].keep_decal_opacity== 1u) {
    alpha = max(alpha, decal_albedo.a);
}
```

Before:
![image](https://github.com/user-attachments/assets/516dc2ed-a7f6-4ace-bdba-b896ddb373af)

After:
![image](https://github.com/user-attachments/assets/c1c717a7-8f1a-46d4-8673-ef5e4ffef6e5)
![image](https://github.com/user-attachments/assets/5738d656-e177-4888-ba75-5d04c0b38c23)

yahkr committed:
Allow decals to modify surface alpha

Calinou commented:
Note that there is no way to opt out of this behavior in this PR. There are some use cases where the decal's opacity should be modulated by the surface opacity, e.g. for blob shadows.

yahkr commented:
> Note that while I agree with the core idea, there's no way to opt out of this behavior in this PR. There are some use cases where the decal's opacity should be modulated by the surface opacity, e.g. for blob shadows.
> 
> I suggest adding a boolean property to Decal that controls this behavior (**Override Surface Alpha** defaulting to `false`).

Added! (much more involved that I anticipated)
![image](https://github.com/user-attachments/assets/ecfab890-95b8-473d-b172-a38cb06cf208)

One thing I feel is important to note, as I'm no graphics expert, I added padding to maintain the correct buffer padding as without it I experienced many graphical issues 
```diff
decal_data_inc.glsl
+ uint override_surface_alpha;
+ uint _pad0;
+ uint _pad1;
+ uint _pad2;

servers/rendering/renderer_rd/storage_rd/texture_storage.h
+ uint32_t override_surface_alpha;
+ uint32_t _pad0;
+ uint32_t _pad1;
+ uint32_t _pad2;
```

Calinou commented:
> One thing I feel is important to note, as I'm no graphics expert, I added padding to maintain the correct buffer size as without it I experienced many graphical issues

This looks correct 🙂 We generally name those padding variables without a leading underscore though, i.e. `pad0`, `pad1` and `pad2`.

Mickeon commented on doc/classes/Decal.xml:
```diff
@@ -87,6 +87,9 @@
 			[b]Note:[/b] Setting [member normal_fade] to a value greater than [code]0.0[/code] has a small performance cost due to the added normal angle computations.
 		</member>
+		<member name="override_surface_alpha" type="bool" setter="set_override_surface_alpha" getter="is_override_surface_alpha_enabled" default="false">
+			If [code]true[/code], the decal's alpha value will replace the surface's alpha wherever the decal's is greater.
```
Assuming this is accurate:
```suggestion
			If [code]true[/code], the alpha on the decal's [member texture_albedo] will replace the surface's alpha onto any pixel where the decal's alpha is greater. This results in the decal always appearing more opaque than the surface.
```
Or similar. It's a bit tricky. It could also use a slightly more elaborate use-case, albeit this is the expected behaviour of decals for most.

yahkr replied:
> If [code]true[/code], the alpha from the decal's [member texture_albedo] will overwrite the surface's alpha for any pixel where the decal's alpha is greater. This ensures that the resulting pixel is at least as opaque as the decal, never less.

Thoughts on this? I feel the "always appearing more opaque" might not be precise enough when describing the edges of a decal that are transparent.

yahkr replied:
possibly an additional sentence talking about how enabling this will give you the same expected behavior that decals have on opaque surfaces?

Mickeon replied:
> If [code]true[/code], the alpha from the decal's [member texture_albedo] will overwrite the surface's alpha for any pixel where the decal's alpha is greater. This ensures that the resulting pixel is at least as opaque as the decal, never less.

"Override" is more appropriate in this case than "overwrite".

```suggestion
			If [code]true[/code], the alpha from the decal's [member texture_albedo] will override the surface's alpha for any pixel where the decal's alpha is greater. This ensures that the resulting pixels are at least as opaque as the decal, never less.
```

> possibly an additional sentence talking about how enabling this will give you the same expected behavior that decals have on opaque surfaces?

Sounds a bit redundant.

Mickeon commented:
This is likely the most expected behaviour for decals, but I'm wondering if the opposite should be available. That is, lowering the alpha of surface with decals. I can't think of a common use-case, so it's probably not worth pursuing.

yahkr commented:
> This is likely the most expected behaviour for decals, but I'm wondering if the opposite should be available. That is, lowering the alpha of surface with decals. I can't think of a common use-case, so it's probably not worth pursuing.

@Mickeon I was thinking that too, and perhaps an additional toggle could be added in the future if a good use-case comes up. The closest i've got would be a decal thats fully opaque with a transparent center.  However this would only apply to transparent surfaces to provide a cutout effect. 

![New Project](https://github.com/user-attachments/assets/0e556dbe-d03a-49c2-a178-509b58631b5f)

![image](https://github.com/user-attachments/assets/6a95b7fd-6dfb-4207-876d-1099b89673f0)

These would need to be separate options, since in cases like this, you really don't want the decal to be able to lower the alpha:

![image](https://github.com/user-attachments/assets/9ec719d5-bcbc-40cc-8578-7fde0fc709fb)
or to just take the decals alpha
![image](https://github.com/user-attachments/assets/940a192d-002f-4279-8684-e9be2c0c7cc0)

This is what it could look like if we add a second option to take the min(alpha, decal.albedo.a) instead:

![image](https://github.com/user-attachments/assets/fa8f6c5d-035e-480b-883e-ea15ce34c1a4)

and of course this only works on transparent materials.
So, while it's technically possible, I agree it's not generally useful and could be confusing unless there's a clear use-case.

Mickeon commented:
I like `keep_decal_opacity` personally.

> Edit: tomorrow I'll swap this to a dropdown and see how that feels

I would suggest not going ahead with this unless there's noticeable demand. Even then, the property can be replaced into an enum property in the future, if desired.

yahkr commented:
@Mickeon 
- changed to `keep_decal_opacity`
- moved shader check above `decal_albedo.a *= fade` (fixes decal being slightly transparent when it shouldn't be)
- updated documentation

MJacred commented:
> This is likely the most expected behaviour for decals, but I'm wondering if the opposite should be available. That is, lowering the alpha of surface with decals. I can't think of a common use-case, so it's probably not worth pursuing.

use cases:
* shooting games: shoot a hole in sth. w/o changing the mesh. IIRC, most people draw to texture to achieve this
* puzzle games: I saw in a detective series that you can look through frosted glass by putting transparent tape over it. So this feature could interest someone doing puzzles, where they want to hide sth. and then give the player a tool to make sth. transparent in a limited manner.

yahkr commented:
updated to 7716a4cba0e4d96e170488bd4f332d22395f7bdb
